crates/executor/src/executor.rs

```diff
@@ -17,7 +18,157 @@ use revm::{
 use std::collections::BTreeMap;
 
 /// Main block executor
-pub struct Executor {}
+pub struct Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    chain_spec: &'a ChainSpec,
+    evm: EVM<&'a mut SubState<DB>>,
+}
+
+impl<'a, DB> Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn new(chain_spec: &'a ChainSpec, db: &'a mut SubState<DB>) -> Self {
+        let mut evm = EVM::new();
+        evm.database(db);
+        Executor { chain_spec, evm }
+    }
+
+    fn recover_signers(
+        &self,
+        body: &[TransactionSigned],
+        signers: Option<Vec<Address>>,
+    ) -> Result<Vec<Address>, Error> {
+        if let Some(signers) = signers {
+            if body.len() == signers.len() {
+                Ok(signers)
+            } else {
+                Err(Error::SignerRecoveryError)
+            }
+        } else {
+            body.iter().map(|tx| tx.recover_signer().ok_or(Error::SignerRecoveryError)).collect()
+        }
+    }
+
+    fn init_block_env(&mut self, header: &Header) {
+        let spec_id = revm_spec(self.chain_spec, header.number);
+        self.evm.env.cfg.chain_id = U256::from(self.chain_spec.chain().id());
+        self.evm.env.cfg.spec_id = spec_id;
+        self.evm.env.cfg.perf_all_precompiles_have_balance = false;
+        self.evm.env.cfg.perf_analyse_created_bytecodes = AnalysisKind::Raw;
+
+        revm_wrap::fill_block_env(&mut self.evm.env.block, header, spec_id >= SpecId::MERGE);
+    }
+}
+
+impl<'a, DB> BlockExecutor for Executor<'a, DB>
+where
+    DB: StateProvider,
+{
+    fn execute(
+        &mut self,
+        block: &Block,
+        signers: Option<Vec<Address>>,
+    ) -> Result<ExecutionResult, Error> {
+        let Block { header, body, ommers } = block;
+        let signers = self.recover_signers(body, signers)?;
```

Comment:
It should be possible to look up the recovered senders using the `StateProvider`. The recovered senders are in the `TxSenders` table

Reply:
I love this idea but also have concern on performance. In `stage` case, the `sender` might be in `TxSenders` but in general case the `sender` might not be in `TxSenders`. Even though we already know it doesn't exist, we'll try to search it first. Doesn't it impact on performance? If I misunderstood, please point me out.

Reply:
Another challenges that I've learned,

* StateProvider doesn't have an interface for TxSenders so I think we need to define and implement `TransactionProvider`.
* `TxSenders`'s key is a (global) tx index. In order to get this tx index, we need to access `BlockBodies` table.

Reply:
@onbjerg Even though I've failed to revise it, I'd like to wrap it up as it is. @gakonst advised me that it'll be revised more internally. Please could you agree on me? If you have any other advice or feedback, please let me know.